### PR TITLE
ROX-16402: Add converters for Matcher types

### PR DESCRIPTION
**Pull request by jvdm**
## Description

1. Introduce type conversion for Scanner Matcher protos. Take into account that conversion from ClairCore to v4 Proto and v4 Proto to ClairCore might be needed to scan vulnerabilities for index reports imported through the `Contents` property of the `GetVulnerabilityReportRequest` proto (use case related to secured cluster scans).
2. Refactored the converters to their package under `service/`.

The Matcher Service will use these types conversions.

| master
| .... :arrow_double_up: **ROX-16402: Add converters for Matcher types #7823** :point_left: 
| ........ :arrow_double_up: fix(scanner): Fix indexer auth when fetching layers #7842
| ............ :arrow_double_up: ROX-16402: Introduce the initial Scanner Matcher #7843

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added

### N/A

- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

## Testing Performed

Added UTs. Verified coverage.

Local run using `scannerctl`:

```
% make -C scanner scanner-go-build 
...
% scanner/image/scanner/bin/scanner -conf scanner-config.yaml
...
```

Followed by:

```
% go run scanner/cmd/scannerctl/main.go -certs ~/tmp/scanner-tls https://registry.access.redhat.com/ubi8/ubi-minimal:latest 
```

Configuration used:

```yaml
indexer:
  db_conn_string: postgresql:///jvdm?host=/var/run/postgresql
matcher:
  db_conn_string: postgresql:///jvdm?host=/var/run/postgresql
mtls:
  certs_dir: /home/jvdm/tmp/scanner-tls
```

### CODE REVIEW

**Review by daynewlee — APPROVED** [review]
Looks good to me. But it would be great if you can get another approval before merging.

File: scanner/services/converters/converters.go
Comment (RTann) [inline scanner/services/converters/converters.go:203]: same here

File: scanner/services/converters/converters.go
Comment (RTann) [inline scanner/services/converters/converters.go:384]: nit: can we give this a more generic name like `convertToMapFunc`? The function doesn't involve ClairCore (just used for that purpose)

File: scanner/services/converters/converters.go
Comment (RTann) [inline scanner/services/converters/converters.go:335]: nit: singular `Distribution`

File: scanner/services/converters/converters.go
Comment (RTann) [inline scanner/services/converters/converters.go:308]: In the other `toSourcePackage`, you use `p.Source`. I think that was mistake, and this one is correct just using `p` here?

File: scanner/services/converters/converters.go
Comment (jvdm) [reply scanner/services/converters/converters.go]: There is a typo, which is noticeable to some extent.

File: scanner/services/converters/converters.go
Comment (jvdm) [reply scanner/services/converters/converters.go:192]: Type conversion is like cloning.

File: scanner/services/converters/converters.go
Comment (jvdm) [reply scanner/services/converters/converters.go:249]: > Want to just grab the first one?

Should we grab the first one? Perhaps the clients can handle it themselves or have the option to use all of them at a later time. Maybe it's not necessary for the Scanner to be concerned about this.

File: scanner/services/converters/converters.go
Comment (jvdm) [reply scanner/services/converters/converters.go:265]: Give an example of how what you're proposing could be helpful. It will help clarify what you mean to everybody, and we can make a decision faster.

Comment (RTann) [reply scanner/services/converters/converters.go:265]: Looked into it, and it won't really work for this context, as we cannot control `claircore.Severity`. If we could, I would have suggested a `Sentinel` severity to go at the end of the enum specified in https://github.com/quay/claircore/blob/main/severity.go#L11, and we use its value to ensure `len(severityMapping) == claircore.Sentinel`. This way, we ensure we cover all severities ever specified in `claircore.Severity`. Something like this:

```
func init() {
        if len(severityMapping) != claircore.Sentinel {
                panic("idk")
        }
}
```

We cannot really do this now, as we cannot really guarantee there will be no severity greater than `Critical` ever defined, so we can't rely check `len(severityMapping) == 1+claircore.Critical`

File: scanner/services/converters/converters.go
Comment (jvdm) [reply scanner/services/converters/converters.go:192]: When mapping or converting structs, we clone them to ensure that the two objects have no standard modifiable references. For read-only references this is fine.

The reason why cloning structs is important when mapping or converting them is to ensure that the two objects have no modifiable references. This helps prevent unintentional modifications to the original object and ensures data integrity. It's especially important when dealing with sensitive information or when multiple processes are accessing the same data. Sometimes objects originally don't have that scope, but later on, they do, and most of the time the human doing the updates will miss the fact that there are common modifiable references between the two objects, and the bug is hard to uncover.

File: scanner/services/converters/converters.go
Comment (jvdm) [reply scanner/services/converters/converters.go:265]: You haven't clarified what you're looking for and what you want to improve. Assuming I understood what your goal is, here's a reference to something that might be related: https://github.com/stackrox/stackrox/blob/jvdm%2FROX-16402%2Fmatcher/scanner/services/converters/converters_test.go#L654

File: scanner/services/converters/converters.go
Comment (dcaravel) [inline scanner/services/converters/converters.go:34]: Should the `hash`/`hashId` be carried over as well?

File: scanner/services/converters/converters.go
Comment (dcaravel) [inline scanner/services/converters/converters.go:136]: This seems risky / may create some tough to troubleshoot bugs in the future, perhaps its worth adding a disclaimer to the function's docs stating the input will be modified.  Or perhaps it's worth creating a function that performs the conversion but ignores the source field?

File: scanner/services/converters/converters.go
Comment (dcaravel) [reply scanner/services/converters/converters.go:265]: Perhaps adding a log / metric when an unmapped severity is encountered will help us identify when this occurs so that we can add the missing mapping.

File: scanner/services/converters/converters.go
Comment (dcaravel) [inline scanner/services/converters/converters.go:306]: Same comment as above regarding converter modifying the input object - perhaps should avoid if possible.

File: scanner/services/converters/converters.go
Comment (jvdm) [reply scanner/services/converters/converters.go:34]: No, we don't use the internal HashID from ClairCore. Instead, we use our own HashID that includes the resource type as a namespace. We plan to scan things besides container images, like nodes.

File: scanner/services/converters/converters.go
Comment (jvdm) [reply scanner/services/converters/converters.go:136]: I agree this is cutting a corner, but can you clarify a scenario where this sanitation might cause problems? Asking because by definition, source packages should not have sources themselves, unless we are talking about some broader interpretation of source packages for non-OS language packages?

Comment (dcaravel) [reply scanner/services/converters/converters.go:136]: I wasn't commenting on data specific scenario (I do not know the data well enough).  The scenario may not exist now for all I know, but if it happens to be valid in the future - this could potentially lead to a difficult bug.

Conventionally I wouldn't expect `clairReport` to be modified in any way if this was executed:
```go
acsVulnReport := converters.ToProtoV4VulnerabilityReport(clairReport)
```

Comment (jvdm) [reply scanner/services/converters/converters.go:136]: ClairCore's response to `Package.Source.Source`:

> we shouldn't ever have a source-source, correct

So sanitizing input here seems reasonable unless you want to block `Package.Source.Source != nil` as invalid input. Can you clarify?

Comment (dcaravel) [reply scanner/services/converters/converters.go:143]: OK, thanks for clarifying - if we're confident this statement will hold true for the 'foreseeable future', then am OK with what you have.  If your extremely confident then could remove the `nil` assignment all together 😉 , but I'd rather see the defensive approach you have here vs. a crash.

My opinion (which will NOT hold up my approval here) is that we shouldn't ever modify the input object in a converter such as this - either populate the source without using recursion (preferred) or reject the input.

Comment (jvdm) [reply scanner/services/converters/converters.go:143]: Maybe the method should be named "Sanitizers" instead. My goal is to **validate and clean** input that I believe is worth cleaning so subsequent logic can work smoothly and abort (with an error) if something is not as expected.

Would you like to change the package and method names to "Sanitize"?

Comment (jvdm) [reply scanner/services/converters/converters.go:143]: Or do you prefer we should block the request altogether?

Comment (dcaravel) [reply scanner/services/converters/converters.go:143]: Yes would rather see that (I know it's a trivial nuance) or clarify in the package docs that input may be modified.

Comment (jvdm) [reply scanner/services/converters/converters.go]: eb18c809db..7d8ddb9c0c

**Review by dcaravel — APPROVED** [review]

File: scanner/services/converters/converters.go
Comment (RTann) [reply scanner/services/converters/converters.go:286]: what's in the test is almost what I was originally looking for. The check that `Critical == 5` works if there are any severities added which are less severe than `Critical` but misses something like `SuperCritical` which would be more severe and unaccounted for

Since we cannot control the severities from here, the best we can do is log it as a warning, which you did

**Review by RTann — APPROVED** [review]
Some minor nits to consider. Otherwise LGTM and no need for another round

**Pull request merged**
